ops/relation.py

```diff
@@ -0,0 +1,548 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import semantic_version as semver
+
+from ops.charm import CharmEvents
+from ops.framework import (
+    Object, EventSource, EventBase, StoredState
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Provider(Object):
+    """Manages relations of resource providers
+
+    A :class:`Provider` object manages relations of a resource provider
+    charms, with charms that consume those resources. When a consumer
+    charm joins the relation a provider object informs it of the type
+    and version of the resource provided. Providers may also be
+    instrumented to inform consumer charms of any relavent
+    information, for example configuration settings. This is done
+    using the `provides` argument of the :class:`Provider` constructor.
+
+    Any charm that provides a resource may choose to do so through the
+    :class:`Provider` object simpliy by instantiating it in the charm's
+    `__init__` method, as follows
+
+    Example::
+
+        self.provider = Provider(self, relation_name, provides)
+        ...
+        self.provider.ready()
+
+    It is important to invoke `ready()` on the :class:`Provider`
+    object, in order to let the consumer charm know that the provider
+    is serving requests. This is done by setting a boolean flag
+    `ready` in the data forwarded to the consumer charm. A provider
+    charm may toggle this flag by invoking `unready()` when it
+    is unable to service any requests for example during an
+    upgrade. After an upgrade the :class:`Provider` object notifies
+    consumer charms by resending type, version and any other data to
+    the consumer. Even though :class:`Provider` objects only handle
+    relation joined and provider upgrade events, they may be
+    subclassed to extend their functionality in any way desired.
+
+    Args:
+        charm: :class:`ops.charm.CharmBase` derived object that is
+            instantiating :class:`Provider`. This is almost always
+            `self`.
+        name: string name of relation (as defined in `metadata.yaml`) that
+            consumer charms will use to interact with provider charms.
+        provides: A dictionary containing the information that
+            :class:`Provider` will forward to any consumer charm when it forms
+            a relation with the provider charm. This dictionary must have at
+            least one key `provides`. The value of this key itself must be a
+            dictionary, whose key as a resource name and value is a version
+            string. For example here is a suitable provides argument
+            Example::
+
+                {'provides': {'mongodb': '2.4'}}
+
+    The `provides` argument dictionary may contain any other key value
+    pairs. If so this information is forwarded to the consumer charm
+    as application data.
+
+    """
+    stored = StoredState()
```

Comment:
usually we've left stored as '_stored' because it shouldn't be part of the public contract with other classes.

Reply:
Agreed. Will fix.